### ISSUE TITLE
x86 base images on Docker Hub are being replaced with ARM base images

### ISSUE THREAD

**Issue by kini** [aws/aws-lambda-base-images#24]
As of this writing, the latest 3.8 and 3.9 images on Docker Hub for amazon/aws-lambda-python are no longer of the linux/amd64 architecture, but linux/arm64 instead. That includes the following versions:

- [3.8.2021.09.29.18 (aafa05498a66)](https://hub.docker.com/layers/amazon/aws-lambda-python/3.8.2021.09.29.18/images/sha256-aafa05498a66bc4decfa83a523d11ad412c2ca4d01a2d2dc3336dc42897d9688?context=explore)
- [3.9.2021.09.29.18 (9c34055a4aa0)](https://hub.docker.com/layers/amazon/aws-lambda-python/3.9.2021.09.29.18/images/sha256-9c34055a4aa09019ae3ec640877ed3ed54bfc8dd1dca93e4f5ae2832fb183576?context=explore)
- [3.8.2021.09.29.19 (57c6e8f2c26b)](https://hub.docker.com/layers/amazon/aws-lambda-python/3.8.2021.09.29.19/images/sha256-57c6e8f2c26b7a117e6e3a063de9529ebad2dd4303cc806eb8e449a92e9bd1b3?context=explore)
- [3.9.2021.09.29.19 (185fb8d8c5fe)](https://hub.docker.com/layers/amazon/aws-lambda-python/3.9.2021.09.29.19/images/sha256-185fb8d8c5fe152135c4729f472bac7b7caee0515d0320e746bdae3c0c782a36?context=explore)
- [3.8 (904f5b61e0e3)](https://hub.docker.com/layers/amazon/aws-lambda-python/3.8/images/sha256-904f5b61e0e3303d9fab2d480ef028ee0aec368eca80c70a04a0ce34a67c1a6f?context=explore)
- [3.8.2021.09.29.20 (904f5b61e0e3)](https://hub.docker.com/layers/amazon/aws-lambda-python/3.8.2021.09.29.20/images/sha256-904f5b61e0e3303d9fab2d480ef028ee0aec368eca80c70a04a0ce34a67c1a6f?context=explore)
- [latest (8db778dbbcca)](https://hub.docker.com/layers/amazon/aws-lambda-python/latest/images/sha256-8db778dbbccab731dc2154d0e88d905fb6b034ec1e25fccfc8edc50744b13e71?context=explore)
- [3 (8db778dbbcca)](https://hub.docker.com/layers/amazon/aws-lambda-python/3/images/sha256-8db778dbbccab731dc2154d0e88d905fb6b034ec1e25fccfc8edc50744b13e71?context=explore)
- [3.9 (8db778dbbcca)](https://hub.docker.com/layers/amazon/aws-lambda-python/3.9/images/sha256-8db778dbbccab731dc2154d0e88d905fb6b034ec1e25fccfc8edc50744b13e71?context=explore)
- [3.9.2021.09.29.20 (8db778dbbcca)](https://hub.docker.com/layers/amazon/aws-lambda-python/3.9.2021.09.29.20/images/sha256-8db778dbbccab731dc2154d0e88d905fb6b034ec1e25fccfc8edc50744b13e71?context=explore)

This has broken my Dockerfile which pulls `FROM amazon/aws-lambda-python:3.8` on a linux/amd64 host. If this is intentional, please let me know how I should be writing my Dockerfile so that it doesn't break because of the new ARM images. Thanks!

EDIT: @josherickson points out in the replies that the same issue exists for the aws-lambda-nodejs image on Docker Hub. The following versions are currently affected:

- [12.2021.09.29.18 (6b70b0dcb576)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/12.2021.09.29.18/images/sha256-6b70b0dcb576add99d2f15ca6d55ebd509f459608be524d8b9ce9f47fae7aa2d?context=explore)
- [14.2021.09.29.18 (e8fa278d8dba)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/14.2021.09.29.18/images/sha256-e8fa278d8dba5cbe31465ad6e3a1aee9bfa2a5cc85ad2c05373e232498e85bf4?context=explore)
- [12.2021.09.29.19 (3b8264fe8ccd)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/12.2021.09.29.19/images/sha256-3b8264fe8ccda5fb30bf9c70928a58d3cb3a3d3482afffc030bbc05fbce13e31?context=explore)
- [14.2021.09.29.19 (ca7eb6997f8b)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/14.2021.09.29.19/images/sha256-ca7eb6997f8b23ace0a4817b924712a22782be92bf0819c0d649d6bec533e0e4?context=explore)
- [latest (082f4ab207a6)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/latest/images/sha256-082f4ab207a691cb7f39cd5b8dc66204e9455f15749cd734097a5e7d2970d843?context=explore)
- [14 (082f4ab207a6)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/14/images/sha256-082f4ab207a691cb7f39cd5b8dc66204e9455f15749cd734097a5e7d2970d843?context=explore)
- [14.2021.09.29.20 (082f4ab207a6)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/14.2021.09.29.20/images/sha256-082f4ab207a691cb7f39cd5b8dc66204e9455f15749cd734097a5e7d2970d843?context=explore)
- [12 (e94c6b086043)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/12/images/sha256-e94c6b08604388b12cb4c457ca2d14aff99f1e96f00179b91429807301e9db82?context=explore)
- [12.2021.09.29.20 (e94c6b086043)](https://hub.docker.com/layers/amazon/aws-lambda-nodejs/12.2021.09.29.20/images/sha256-e94c6b08604388b12cb4c457ca2d14aff99f1e96f00179b91429807301e9db82?context=explore)

It does look like things are fine on ECR Public, though (?).

**Comment by vit-spinka-hvr** [comment]
I don't think this is intentional. Docker build (with buildkit disabled) says explicitly. (I even specified --platform to force amd64.)

```
WARNING: Pulled image with specified platform (linux/amd64), but the resulting image's configured platform (linux/arm64/v8) does not match.
This is most likely caused by a bug in the build system that created the fetched image (amazon/aws-lambda-python:3.9.2021.09.29.20).
Please notify the image author to correct the configuration.
```

**Comment by kini** [comment]
Yes, I also don't think it's intentional, but one never knows...

Here's the workaround I'm using for now:

```
FROM amazon/aws-lambda-python:3.8.2021.09.13.11
```

**Comment by josherickson** [comment]
Same thing for the nodejs images. For anyone getting here, you can check this list here and go back to find a previous version that's still using `linux/amd64`.

https://hub.docker.com/r/amazon/aws-lambda-nodejs/tags?page=1&name=14

The symptom we were seeing is the following during our Docker build when running something like a `RUN yum ...` or `RUN curl ...` command. I'm guessing you'd get this with any `RUN` statement in your Dockerfile, but haven't tested it any further.
```
standard_init_linux.go:228: exec user process caused: exec format error
```

**Comment by kini** [comment]
Thanks, I've updated the OP to include info about the broken nodejs image versions. Here are workaround FROM lines for node, which I assume should work, based on the current latest amd64 image for 12 and 14:

```
FROM amazon/aws-lambda-nodejs:12.2021.09.13.11
```

```
FROM amazon/aws-lambda-nodejs:14.2021.09.13.11
```

**Comment by Tenzer** [comment]
It could look like the same Docker images in the ECR registry supports both architectures, so `public.ecr.aws/lambda/python:3.9` and `public.ecr.aws/lambda/nodejs:14` for instance work correctly.

**Comment by jakemraz** [comment]
Same here. I use Mac M1 and got a same error on a python 3.8 image.

**Comment by anentropic** [comment]
Wouldn't it be better to publish multi-arch images?

Currently the `3.11` tag points only to amd64, there is no arm64 image for the `3.11` tag

If I look at todays builds I can see distinct datetime tags on amd64 and arm64 images:
https://hub.docker.com/r/amazon/aws-lambda-python/tags?page=1&name=3.11.2023.11.07

perhaps this is only because they are an hour apart (judging by the tag)?

given the rest of the issue above... is it that both arch images are published separately and then there's a race to see which one gets the tag, for the less granular tags?

that's why it would be better to publish proper multi-arch images - then the same tag can apply to images for both archs.

A multi-arch image looks like this:
https://hub.docker.com/layers/anentropic/chuckd/0.5.4/images/sha256-db261bc0a3201f9485b67f2b65ce2e91cb7e2e1fe1a046cf678e2e91c07ab900?context=repo

...a single tag with options in the "OS/ARCH" drop down, and as a user our Docker cli will just pull the appropriate one automatically.

It's quite sad this issue persists for 2 years with no attention 😞

**Comment by anentropic** [comment]
presumably intended solution is to use ECR instead of Docker Hub

the images on ECR have arch suffix in the tags, e.g.:
[public.ecr.aws/lambda/python:3.11-arm64](https://gallery.ecr.aws/lambda/python)

that would also be a good solution to apply to the Docker Hub images if building multi-arch images is too hard
